File: marble-backoffice/src/models/CredentialsDto.ts
import * as yup from "yup";
import { adaptDtoWithYup } from "@/infra/adaptDtoWithYup";
import { adaptRole } from "./Role";
import { Credentials } from "./Credentials";

const CredentialsSchema = yup.object({
  organization_id: yup.string().defined(),
  role: yup.string().required(),
  actor_identity: yup
    .object({
      user_id: yup.string(),
      email: yup.string(),
      api_key_name: yup.string(),
    })
    .required(),
});

export function adaptCredential(json: unknown): Credentials {

  const dto = adaptDtoWithYup(
    json,
    yup.object({
      credentials: CredentialsSchema,
    })
  ).credentials

  return {
    organizationId: dto.organization_id,
    role: adaptRole(dto.role),
    actorIdentity: {
        userId: dto.actor_identity.user_id || "",
        email: dto.actor_identity.email || "",
        apiKeyName: dto.actor_identity.api_key_name || "",
    },
  };
}

File: server/dto/credentials_dto.go
package dto

import "marble/marble-backend/models"

type Identity struct {
	UserId     string `json:"user_id,omitempty"`
	Email      string `json:"email,omitempty"`
	ApiKeyName string `json:"api_key_name,omitempty"`
}

type Credentials struct {
	OrganizationId string   `json:"organization_id"`
	Role           string   `json:"role"`
	ActorIdentity  Identity `json:"actor_identity"`
}

func AdaptCredentialDto(creds models.Credentials) Credentials {
	return Credentials{
		OrganizationId: creds.OrganizationId,
		Role:           creds.Role.String(),
		ActorIdentity: Identity{
			UserId:     string(creds.ActorIdentity.UserId),
			Email:      creds.ActorIdentity.Email,
			ApiKeyName: creds.ActorIdentity.ApiKeyName,
		},
	}
}

func AdaptCredential(dto Credentials) models.Credentials {
	return models.Credentials{
		OrganizationId: dto.OrganizationId,
		Role:           models.RoleFromString(dto.Role),
		ActorIdentity: models.Identity{
			UserId:     models.UserId(dto.ActorIdentity.UserId),
			Email:      dto.ActorIdentity.Email,
			ApiKeyName: dto.ActorIdentity.ApiKeyName,
		},
	}
}
